File: Deep-Learning-for-Beginners-master/R/ch.3/TestBackpropCE.R
source("./BackpropCE.R")
source("./Sigmoid.R")

X <- matrix(c(
  0, 0, 1,
  0, 1, 1,
  1, 0, 1,
  1, 1, 1
), nrow=4, ncol=3, byrow=TRUE)

D <- c(
  0,
  1,
  1,
  0
)

W1 <- matrix(runif(12, min=-1, max=1), nrow=4, ncol=3)
W2 <- matrix(runif(4, min=-1, max=1), nrow=1, ncol=4)

for (epoch in 1:10000) { # train
  Ws <- BackpropCE(W1, W2, X, D)
  W1 <- Ws$W1
  W2 <- Ws$W2
}

N <- 4 # inference
for (k in 1:N) {
  x <- X[k,]
  v1 <- W1 %*% x
  y1 <- Sigmoid(v1)
  v <- W2 %*% y1
  y <- Sigmoid(v)
  print(y)
}

File: Deep-Learning-for-Beginners-master/Matlab/ch.3/TestBackpropCE.m
% 该文件对 BackpropCE 函数进行了测试。

clear all

X = [ 0 0 1;
      0 1 1;
      1 0 1;
      1 1 1;
    ];

D = [ 0
      1
      1
      0
    ]; % XOR 问题

W1 = 2*rand(4, 3) - 1; % 随机初始化权重参数
                       % 每个元素的值在 -1 到 1 之间
W2 = 2*rand(1, 4) - 1; % 随机初始化权重参数
                       % 每个元素的值在 -1 到 1 之间

% 训练模型
for epoch = 1:10000
  [W1 W2] = BackpropCE(W1, W2, X, D);
end

% 使用模型进行预测
N = 4;
for k = 1:N
  x  = X(k, :)';
  v1 = W1*x;
  y1 = Sigmoid(v1);
  v  = W2*y1;
  y  = Sigmoid(v)
end
